### DISCUSSION TITLE
Writing hot cue and memory cue points

### DISCUSSION THREAD

**Discussion by laker-93** [dylanljones/pyrekordbox#113] [General]
Hello and Happy New Year! Many thanks for this awesome project.
I was wondering what the plan is for being able to write both memory cue points and hot cue points to the meta data files?
The readme currently says:

"The files have names like ANLZ0000 and come with the extensions .DAT, .EXT or .2EX. They include waveforms, beat grids (information about the precise time at which each beat occurs), time indices to allow efficient seeking to specific positions inside variable bit-rate audio streams, and lists of memory cues and loop point."

So as it stands, is pyrekordbox able to parse the files were memory cue points are set? What needs to be done in order to write memory and hot cue points? I'd be happy to work on this.

Thanks,

Luke

**Comment by dylanljones** [comment]
Hey Luke,

Thank you and a happy new year to you too!

Cue points in Rekordbox are a bit confusing. The local cue points are stored in the DjmdCues table of the master.db database, not in the ANLZ files. I am guessing the cue points are written to the *exported* ANLZ files when preparing the USB drive. I haven't checked the exported ANLZ files though, so far I only looked at the local ANLZ files. In the exported ANLZ files, the cue points *should* be stored in the `PCOB` and `PCO2` tags. These tags are empty in the local ANLZ files for me.

Accessing the cue points in the database is really straight forward. Using the newest version `v0.3.2`, the cue points of a `DjmdContent` can be accessed via a simple relationship:

````python
from pyrekordbox import Rekordbox6Database

db = Rekordbox6Database()
content = db.get_content(ID=<content id>)
for cue in content.Cues:
    print(cue)
````

I actually just started to work on full cue point support in the database handler, check the cue branch [here](https://github.com/dylanljones/pyrekordbox/blob/9cff87545653885e8c32d2bf72dd5160b071d936/pyrekordbox/db6/database.py#L1870-L2150);) Adding new cue points to the database is simple for most tracks, however I am a little stuck on tracks with variable or average bit rate (ABR/VBR). For this the `InMpegFrame` and `InMpegAbs` are required. The latter is the byte offset in the file where the starting frame appears (see [djmdcue](https://pyrekordbox.readthedocs.io/en/latest/formats/db6.html#djmdcue)). Any help on this is more than welcome:)

**Reply by laker-93** [reply]
Thanks for the detailed response. So presumably this targets rekordbox 6 exclusively for now without rekordbox 5 support? Also is the MBR/ABR only required for supporting cue writes to tracks from streaming services? If so perhaps this could be deferred to a separate issue so that the main feature of writing cues for locally stored tracks can be released.

**Reply by dylanljones** [reply]
ABR/VBR is a feature of Mp3 files, which consist of MPEG frames. Each frame *can* have a different bit or sample rate. These values are stored in the frame header, see for example [here](http://mpgedit.org/mpgedit/mpeg_format/mpeghdr.htm). I don't really use streaming services, so I don't know how those tracks are represented in the database, but I guess this also applies to streamed tracks.
At least for Mp3 files it would be necessary to read the frame headers and figure out the bit rate for each frame.

**Reply by dylanljones** [reply]
The MP3 frame headers can now be parse via Ffprobe or a pure Python implementation as fallback, see the `cues` branch.
Checking for VBR/ABR encoding works as expected, but I still can't reproduce the correct `InMpegAbs` values. 
Until we can figure this out, a `NotImplementedError` is raised if a VBR/ABR encoded MP3 file is detected.

**Reply by laker-93** [reply]
Hey Dylan. What's the latest status with the cues branch? Happy to pick this up. For my use case, I don't need streaming track support. Just being able to read/write cue points for normal audio files is good enough for me. Specifically, I'd like to read the cue points from a RB 5 or RB 6 XML, then create a new XML (5 or 6) and write the cue points back in to the new file - thereby preserving the cue information when creating a new XML from an old one. Thanks :)

**Reply by Bide-UK** [reply]
I've been thinking/looking into this problem on off for longer than I'd like to [admit](https://github.com/Bide-UK/Bides-Rekordbox-AutoCuer). 
After the car crash that is Rekordbox 7 and how they're now essentially charging for this feature... I'm also happy to look at contributing.
I've looked at the commit, if you're able to provide an update on what you've implemented/tried that would be a great place to know where to start. Support VBR/ABR is non trivial but if you can support fixed bitrate tracks to begin wit that would a great place to start.

**Reply by laker-93** [reply]
Hey @dylanljones and @Bide-UK - have you guys looked in to this anymore?
On the serato side I've found this very helpful documentation and code to see how to extract and write cue points from the id3 tags that serato writes for its cue points. May be useful to take inspiration from this for rekordbox:

https://github.com/Holzhaus/mixxx/blob/6f8da175fb084440e9f72db90ff83579c47f8637/src/track/serato/tags.cpp#L106
https://github.com/Holzhaus/serato-tags/blob/main/docs/writeup.md

I think as a first effort we can exclude the streaming track use case and "just" try to get read/write cue point support for say mp3 files. We can then add to this with more file support incrementally. What do you think, @dylanljones? Any pointers of how we can pick things up from your cue branch to achieve this?

**Reply by dylanljones** [reply]
Hey @laker-93, sorry for being inactive for so long:/

I have tried quite a few thinks, still stuck on the issue with VBR/ABR encoded files. Files other than MP3s are pretty easy to handle, so the main issue is still MP3. I think we could get it to work if we - for now - exclude VBR support. Even though it would be nice to have that, since there will be a few files encoded like that. All progress i have made is in the `cues` branch if you want to have a go at it.
Also, I haven't looked into streaming files at all since i don't use them. I don't think that's top prio, but if anyone wants to have a look at that go for it:)

As always, any help is appreciated:)
 
Thank you for the links btw, will check them out as soon as the library is back up to date and RB7 support is in!

**Comment by laker-93** [comment]
I've managed to get something working by writing cue points to the xml file. Seems like you just need to tag the track with a TotalTime value in seconds, then you can add POSITION_MARK for cue/memory/loops. Below is a script to do this. It's good enough for my purposes since I don't mind relying on the xml (for now).

```
from pyrekordbox.rbxml import RekordboxXml
from pathlib import Path
import subprocess
import re
 
xml = RekordboxXml()

def get_duration_seconds(path: Path) -> int:
    process = subprocess.Popen(['ffmpeg',  '-i', path], stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
    stdout, stderr = process.communicate()
    matches = re.search(r"Duration:\s{1}(?P<hours>\d+?):(?P<minutes>\d+?):(?P<seconds>\d+\.\d+?),", stdout.decode(), re.DOTALL).groupdict()
    hours = int(round(float(matches['hours'])))
    minutes = int(round(float(matches['minutes'])))
    seconds = int(round(float(matches['seconds'])))
    return int(seconds + (minutes * 60) + (hours * 60 * 60))

path = Path("/Users/lukepurnell/Downloads/Skee Mask - C/Skee Mask - C - 06 One For Vertigo.mp3")
assert path.exists()
track = xml.add_track(path)
duration = get_duration_seconds(path)
track["TotalTime"] = duration
track.add_mark(Name="A", Type="cue", Start=50, Num=1)
track["Name"] = "One For Vertigo"
xml.save('rekordbox.xml')
```

**Reply by Bide-UK** [reply]
Good work and good tune! Do you not also need to change the binary file that's used on the device?

**Reply by laker-93** [reply]
what do you mean by change the binary file on the device? The above is just for creating an xml. You can then import the contents from the xml in to Rekordbox using the xml interface support in the rekordbox app.
